src/databricks/labs/ucx/source_code/notebook.py

```diff
@@ -0,0 +1,331 @@
+        return CellLanguage.MARKDOWN
+
+    def is_runnable(self) -> bool:
+        return True  # TODO
+
+    def build_dependency_graph(self, parent: NotebookDependencyGraph):
+        pass  # not in scope
+
+
+class RunCell(Cell):
+
+    @property
+    def language(self):
+        return CellLanguage.RUN
+
+    def is_runnable(self) -> bool:
+        return True  # TODO
+
+    def build_dependency_graph(self, parent: NotebookDependencyGraph):
+        command = f'{LANGUAGE_PREFIX}{self.language.magic_name}'.strip()
+        lines = self._original_code.split('\n')
+        for line in lines:
+            start = line.index(command)
+            if start >= 0:
+                path = line[start + len(command) :].strip()
+                parent.register_dependency(path.strip('"'))
+                return
+        raise ValueError("Missing notebook path in %run command")
+
+
+class CellLanguage(Enum):
+    # long magic_names must come first to avoid shorter ones being matched
+    PYTHON = Language.PYTHON, 'python', '#', PythonCell
+    SCALA = Language.SCALA, 'scala', '//', ScalaCell
+    SQL = Language.SQL, 'sql', '--', SQLCell
+    RUN = None, 'run', None, RunCell
+    MARKDOWN = None, 'md', None, MarkdownCell
+    R = Language.R, 'r', '#', RCell
+
+    def __init__(self, *args):
+        super().__init__()
+        self._language = args[0]
+        self._magic_name = args[1]
+        self._comment_prefix = args[2]
+        self._new_cell = args[3]
+
+    @property
+    def language(self) -> Language:
+        return self._language
+
+    @property
+    def magic_name(self) -> str:
+        return self._magic_name
+
+    @property
+    def comment_prefix(self) -> str:
+        return self._comment_prefix
+
+    @classmethod
+    def of_language(cls, language: Language) -> CellLanguage:
+        return next((cl for cl in CellLanguage if cl.language == language))
+
+    @classmethod
+    def of_magic_name(cls, magic_name: str) -> CellLanguage | None:
+        return next((cl for cl in CellLanguage if magic_name.startswith(cl.magic_name)), None)
+
+    def read_cell_language(self, lines: list[str]) -> CellLanguage | None:
+        magic_prefix = f'{self.comment_prefix}{MAGIC_PREFIX}'
+        magic_language_prefix = f'{magic_prefix}{LANGUAGE_PREFIX}'
+        for line in lines:
+            # if we find a non-comment then we're done
+            if not line.startswith(self.comment_prefix):
+                return None
+            # if it's not a magic comment, skip it
+            if not line.startswith(magic_prefix):
+                continue
+            if line.startswith(magic_language_prefix):
+                line = line[len(magic_language_prefix) :]
+                return CellLanguage.of_magic_name(line.strip())
+            return None
+        return None
+
+    def new_cell(self, source: str) -> Cell:
+        return self._new_cell(source)
+
+
+def extract_cells(source: str, default_language: CellLanguage) -> list[Cell] | None:
```

Comment:
shouldn't it be a method on `CellLanguage`?

Reply:
I don't think so, it doesn't change from 1 language to another, except for the comment prefix. CellLanguage is really about those variations. But it wouldn't be bad.

Reply:
Second thoughts, I think you're right. I changed it.